### PR TITLE
longest common prefix algorithm

### PR DESCRIPTION
**Pull request by maxhealy01**
## Summary
This is a solution to the Longest Common Prefix challenge. A description can be found below.

## Thought Process
I learned something new about JavaScript here. Using nested loops, I moved one letter at a time, making sure that each string matched the resulting substring. 

However, I needed to break the outer loop based on a condition in the inner loop, so after a bit of research I realized that this can be achieved by naming the outer loop. I did so, and I think the solution is elegant, though I'm not sure if this is a poor choice for readability, given the rarity of the practice.

## Issue
https://github.com/maxhealy01/blind-75/issues/5

## Checklist
- [x] Contains JSDoc comments
- [x] Contains tests
- [x] Completed self review

### CODE REVIEW

File: week_1/String/longest_common_prefix.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+/**
+ * @param {string[]} strs
```
Comment (tannaurus) [inline week_1/String/longest_common_prefix.js:2]: Let's fill this out with a meaningful bit of documentation to describe what this function is designed to do 👍

File: week_1/String/longest_common_prefix.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+ * @return {string}
+ */
+const longestCommonPrefix = (words) => {
```
Comment (tannaurus) [inline week_1/String/longest_common_prefix.js:5]: Can we add some tests? As it stands, I don't have a great way of confirming if the behavior here works or not

File: week_1/String/longest_common_prefix.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+	let longest_common_prefix = "";
+	outer: for (letter = 0; letter < words[0].length; letter++) {
+		for (word = 1; word < words.length; word++) {
```
Comment (tannaurus) [inline week_1/String/longest_common_prefix.js:8]: Without an issue or broader description, I don't know what the goals of this function are. Can we link that in the PR?

**Comment by tannaurus** [comment]
Just a reminder, make sure to fill out the PR checklist before sending it my way 👍

File: week_1/String/longest_common_prefix.js (hunk 1 above)
Comment (maxhealy01) [reply week_1/String/longest_common_prefix.js:5]: I added a description to the beginning of the file.

File: week_1/String/longest_common_prefix.js (hunk 2 above)
Comment (maxhealy01) [reply week_1/String/longest_common_prefix.js:8]: I added the test file, but it's so long that it's difficult to sift through. I think the file needs to be split up. Do you think that having a separate file for each test would be extra/confusing?

Comment (tannaurus) [reply week_1/String/longest_common_prefix.js:8]: Usually you'd have a test file for each logic file. 

So you'd see something like
- `string/longest_common_prefix.js`
- `string/longest_common_prefix.test.js`

both sitting next to each other. It's not uncommon to see them in a test directory like you have here, too.

File: week_1/__tests__/Array.test.js [hunk 4]
```diff
@@ -28,3 +33,35 @@ test("max_subarray works", () => {
...
+	expect(mergeSortedArrays([1, 4, 6, 7, 9], [1, 3, 5, 5, 7, 8, 9, 20])).toEqual(
+		[1, 1, 3, 4, 5, 5, 6, 7, 7, 8, 9, 9, 20]
+	);
+});
+
+test("count_unique_values works", () => {
+	expect(countUniqueValues([[1, 2, 3, 4, 4, 4, 7, 7, 12, 12, 13]])).toEqual(7);
+});
+
+test("max_subarray_sum works", () => {
```
Comment (tannaurus) [inline week_1/__tests__/Array.test.js:51]: instead of "works", describe what "works" means

"max_subarray_sum can take in an array of numbers and return 'something'

This helps a success or failure reflect a specific piece of something succeeding or failing. It might be easier to picture each of these tests as a part of a broader application, maybe as components in a React app. You to use the test name a descriptor for an action that you expect to succeed. "my_function will return an error if given an array of strings", "my_function will sum up all of the _insert_condition_ and return the summed values", etc

File: week_1/__tests__/String.test.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+const romanToInt = require("../String/roman_to_int");
+const {
+	longest_palindrome,
+	longestPalindrome,
+} = require("../String/longest_palindrome");
+const isValidParentheses = require("../String/valid_parentheses");
+const isSubsequence = require("../String/is_subsequence");
+const longestCommonPrefix = require("../String/longest_common_prefix");
+
+test("valid_anagram works", () => {
```
Comment (tannaurus) [inline week_1/__tests__/String.test.js:12]: All these tests are great, but let's keep this MR scoped to something manageable: the prefix algorithm. We can split this diff off into a separate MR and get that merged independently :)

File: week_1/String/longest_common_prefix.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
+// Write a function to find the longest common prefix string amongst an array of strings.
+
+// If there is no common prefix, return an empty string "".
+/**
+ * @param {string[]} strs
```
Comment (tannaurus) [inline week_1/String/longest_common_prefix.js:5]: let's be a bit more descriptive than "strs"
"words that will be checked for a common prefix"

File: week_1/String/longest_common_prefix.js [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+ * @return {string}
```
Comment (tannaurus) [inline week_1/String/longest_common_prefix.js:6]:
```suggestion
 * @return {string} The longest common prefix
```

File: week_1/String/longest_common_prefix.js [hunk 8]
```diff
@@ -0,0 +1,21 @@
+// Write a function to find the longest common prefix string amongst an array of strings.
+
+// If there is no common prefix, return an empty string "".
```
Comment (tannaurus) [inline week_1/String/longest_common_prefix.js:3]: Can we update this to use the jsdoc `example` tag?

https://jsdoc.app/tags-example.html

File: week_1/__tests__/String.test.js [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+	expect(isAnagram("awesome", "awesom")).toBe(false);
+});
+
+test("is_subsequence works", () => {
+	expect(isSubsequence("tart", "trading rats")).toBe(true);
+	expect(isSubsequence("mouthwash", "hsawhtuom")).toBe(false);
+	expect(isSubsequence("hair", "nair")).toBe(false);
+});
+
+test("longest_common_prefix works", () => {
```
Comment (tannaurus) [inline week_1/__tests__/String.test.js:62]: Let's use a describe here

```js
describe("longest_common_prefix", () => {
  test("will find prefix when all words share the same prefix", () => {
    expect(longestCommonPrefix(["flower", "flow", "flytrap"])).toBe("fl")
  })
}
```

File: week_1/__tests__/String.test.js [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+	expect(longestCommonPrefix(["flower", "flow", "flytrap"])).toBe("fl");
+	expect(longestCommonPrefix(["oh no!", "pies", "peanuts"])).toBe("");
```
Comment (tannaurus) [inline week_1/__tests__/String.test.js:64]: Will this test pass too?

```js
expect(longestCommonPrefix(["pies", "peanuts", "oh no!"])).toBe("");
```

File: week_1/String/longest_common_prefix.js [hunk 11]
```diff
@@ -0,0 +1,21 @@
+// Write a function to find the longest common prefix string amongst an array of strings.
```
Comment (tannaurus) [inline week_1/String/longest_common_prefix.js:1]: Looks like this is a duplicate file now?

Comment (maxhealy01) [reply week_1/String/longest_common_prefix.js]: I deleted it

File: week_1/__tests__/Array.test.js [hunk 12]
```diff
@@ -28,3 +33,35 @@ test("max_subarray works", () => {
 test("two_sum works", () => {
 	expect(twoSum([2, 7, 11, 15], 9)).toEqual([0, 1]);
 });
+
+test("majority_element works", () => {
```
Comment (tannaurus) [inline week_1/__tests__/Array.test.js:37]: can we remove all of these unrelated tests? everything else looks good 😄

Comment (maxhealy01) [reply week_1/__tests__/Array.test.js]: I think that I removed them. I tried to do it from the "Files Changed" section. Did that work?

**Comment by tannaurus** [comment]
Looks great 👍 nice work!
